crates/accelerate/src/remove_diagonal_gates_before_measure.rs

```diff
@@ -0,0 +1,103 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+/// Remove diagonal gates (including diagonal 2Q gates) before a measurement.
+use pyo3::prelude::*;
+use std::collections::HashSet;
+
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType};
+use qiskit_circuit::operations::Operation;
+use qiskit_circuit::operations::StandardGate;
+
+/// Run the RemoveDiagonalGatesBeforeMeasure pass on `dag`.
+/// Args:
+///     dag (DAGCircuit): the DAG to be optimized.
+/// Returns:
+///     DAGCircuit: the optimized DAG.
+#[pyfunction]
+#[pyo3(name = "remove_diagonal_gates_before_measure")]
+fn run_remove_diagonal_before_measure(dag: &mut DAGCircuit) -> PyResult<()> {
+    let diagonal_1q_gates = HashSet::from([
+        StandardGate::RZGate,
+        StandardGate::ZGate,
+        StandardGate::TGate,
+        StandardGate::SGate,
+        StandardGate::TdgGate,
+        StandardGate::SdgGate,
+        StandardGate::U1Gate,
+        StandardGate::PhaseGate,
+    ]);
+    let diagonal_2q_gates = HashSet::from([
+        StandardGate::CZGate,
+        StandardGate::CRZGate,
+        StandardGate::CU1Gate,
+        StandardGate::RZZGate,
+        StandardGate::CPhaseGate,
+        StandardGate::CSGate,
+        StandardGate::CSdgGate,
+    ]);
+
```

Comment:
Note that there are some 1-qubit and 2-qubit diagonal gates that did not appear in the original list, so I added them here (I'll add them to the tests later).

There is also a 3-qubit diagonal gate: [`CCZGate`](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.CCZGate)

Reply:
In fact, there is also an n-qubit diagonal gate: [`MCPhaseGate`](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.MCPhaseGate). This gate was not handled this PR.
This is since the algorithm given here that goes over all the successors of each of the predecessors would be `O(n^2)` and not `O(n)`.